### PR TITLE
Build coreos grub off of alpine:3.9 and separate build steps for caching

### LINKED ISSUE #3295: coreos grub compile fails with Alpine 3.9 tool chain
While updating the alpine base to Alpine 3.9 the build of `mkimage-iso-efi`, `mkimage-qcow2-efi`, and `mkimage-raw-efi` fails on x86 and arm64 with the following warnings:
```
gcc -DHAVE_CONFIG_H -I.  -Wall -W -DGRUB_UTIL=1 -D_FILE_OFFSET_BITS=64 -I./include -DGRUB_FILE=\"grub-core/osdep/init.c\" -I. -I. -I. -I. -I./include -I./include -I./grub-core/lib/libgcrypt-grub/src/ -I./grub-core/gnulib -I./grub-core/gnulib  -D_FILE_OFFSET_BITS=64 -fno-builtin-gettext
 -Wall -W -Wshadow -Wpointer-arith -Wundef -Wchar-subscripts -Wcomment -Wdeprecated-declarations -Wdisabled-optimization -Wdiv-by-zero -Wfloat-equal -Wformat-extra-args -Wformat-security -Wformat-y2k -Wimplicit -Wimplicit-function-declaration -Wimplicit-int -Wmain -Wmissing-braces -Wmi
ssing-format-attribute -Wmultichar -Wparentheses -Wreturn-type -Wsequence-point -Wshadow -Wsign-compare -Wswitch -Wtrigraphs -Wunknown-pragmas -Wunused -Wunused-function -Wunused-label -Wunused-parameter -Wunused-value  -Wunused-variable -Wwrite-strings -Wnested-externs -Wstrict-protot
ypes -Wcast-align  -Wextra -Wattributes -Wendif-labels -Winit-self -Wint-to-pointer-cast -Winvalid-pch -Wmissing-field-initializers -Wnonnull -Woverflow -Wvla -Wpointer-to-int-cast -Wstrict-aliasing -Wvariadic-macros -Wvolatile-register-var -Wpointer-sign -Wmissing-include-dirs -Wmissi
ng-prototypes -Wmissing-declarations -Wformat=2 -Werror -Wno-undef -Wno-sign-compare -Wno-unused -Wno-unused-parameter -Wno-redundant-decls -Wno-unreachable-code -Wno-conversion  -Os -Wno-unused-value -MT grub-core/osdep/grub_script_check-init.o -MD -MP -MF grub-core/osdep/.deps-util/g
rub_script_check-init.Tpo -c -o grub-core/osdep/grub_script_check-init.o `test -f 'grub-core/osdep/init.c' || echo './'`grub-core/osdep/init.c
mv -f grub-core/osdep/.deps-util/libgrubkern_a-sleep.Tpo grub-core/osdep/.deps-util/libgrubkern_a-sleep.Po
In file included from grub-core/partmap/gpt.c:26:
./include/grub/gpt_partition.h:90:1: error: alignment 1 of 'struct grub_gpt_header' is less than 8 [-Werror=packed-not-aligned]
 } GRUB_PACKED;
 ^
./include/grub/gpt_partition.h:100:1: error: alignment 1 of 'struct grub_gpt_partentry' is less than 8 [-Werror=packed-not-aligned]
 } GRUB_PACKED;
 ^
./include/grub/gpt_partition.h:100:1: error: alignment 1 of 'struct grub_gpt_partentry' is less than 8 [-Werror=packed-not-aligned]
```

when disabling this warning by adding `-Wno-packed-not-aligned` to the `CFLAGS` the compile succeeds on arm64 but still fails on x86 with:

```
 ../build-aux/install-sh -c -d '/usr/local/share/info'
make[3]: Leaving directory '/grub/docs'
make[2]: Leaving directory '/grub/docs'
Making install in util/bash-completion.d
make[2]: Entering directory '/grub/util/bash-completion.d'
make[3]: Entering directory '/grub/util/bash-completion.d'
make[3]: Nothing to be done for 'install-exec-am'.
 ../../build-aux/install-sh -c -d '/usr/local/etc/bash_completion.d'
 /usr/bin/install -c -m 644 grub '/usr/local/etc/bash_completion.d'
make[3]: Leaving directory '/grub/util/bash-completion.d'
make[2]: Leaving directory '/grub/util/bash-completion.d'
make[1]: Leaving directory '/grub'
./grub-mkimage: error: relocation 0x4 is not implemented yet.
The command '/bin/sh -c mkdir /grub-lib &&   set -e &&   git clone https://github.com/coreos/grub.git &&   cd grub &&   git checkout -b grub-build ${GRUB_COMMIT} &&   ./autogen.sh &&   ./configure --libdir=/grub-lib --with-platform=efi CFLAGS="-Os -Wno-unused-value -Wno-packed-not-aligned" &&   make -j "$(getconf _NPROCESSORS_ONLN)" &&   make install &&   case $(uname -m) in   x86_64)     ./grub-mkimage -O x86_64-efi -d /grub-lib/grub/x86_64-efi -o /grub-lib/BOOTX64.EFI -p /EFI/BOOT ${GRUB_MODULES} linuxefi;     ;;   aarch64)     ./grub-mkimage -O arm64-efi -d /grub-lib/grub/arm64-efi -o /grub-lib/BOOTAA64.EFI -p /EFI/BOOT ${GRUB_MODULES};     ;;   esac' returned a non-zero code: 1
exit status 1
```

We should probably move off an unmaintained version of grub.

/cc @deitch @arm64b

### PR DESCRIPTION
**Pull request by deitch**
Signed-off-by: Avi Deitcher <avi@deitcher.net>

fixes #3295 

**- What I did**

* Rebuilt from `alpine:3.9`, which includes `gcc:8.x` (`alpine:3.8` was off of `gcc:6.x`)
* Moved to latest commit of `coreos/grub` from a few months ago that fixes the indentation issue
* Applied a patch directly to fix an issue on `x86_64`
* Updated the dependent tools images: `tools/mkimage-iso-efi`, `tools/mkimage-raw-efi`, `tools/mkimage-qcow2-efi`
* Updated `src/cmd/linuxkit/moby/output.go` to use these latest images

**- How I did it**

A lot of debugging :-)

**- How to verify it**

Run, build, see that outputs are bootable.

**- Description for the changelog**

Update grub to run off of `alpine:3.9`

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by deitch** [comment]
Hmm, CI failed with 3 errors, all from `kmod`:

```
[+ 37m 26s] [FAIL    ] linuxkit.kernel.kmod_4.14.x 13.36s
[+ 37m 38s] [FAIL    ] linuxkit.kernel.kmod_4.19.x 12.33s
[+ 37m 53s] [FAIL    ] linuxkit.kernel.kmod_5.0.x 14.48s
```

Can you look please @rn ?

**Comment by rn** [comment]
The CI failure is expected. The latest kernels are build with gcc 8.3.0 while the test is still using alpine base with gcc 8.2.0. Once the 0.7 stuff is merged and you rebased it should go away

**Comment by deitch** [comment]
> Once the 0.7 stuff is merged and you rebased it should go away

Sounds good.

**Comment by deitch** [comment]
Rebased off of post-0.7 master, CI restarting. @rn once this passes, should be good to go.

**Comment by deitch** [comment]
Excellent! CI is all green

File: tools/grub/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM linuxkit/alpine:3683c9a66cd4da40bd7d6c7da599b2dcd738b559 AS grub-build
+FROM linuxkit/alpine:518c2ed0f398c5508969ac5e033607201fb419ed AS grub-build
```
Comment (rn) [inline tools/grub/Dockerfile]: this should be `linuxkit/alpine:86cd4f51b49fb9a078b50201d892a3c7973d48ec`. The Alpine based was updated earlier this week

Comment (deitch) [reply tools/grub/Dockerfile]: I took the last tag from when I did this, probably Monday, as I flew on Tuesday. It is a bit slow to update it (earliest `FROM` line, causes an entire rebuild), will try to get it done.

Comment (deitch) [reply tools/grub/Dockerfile]: Updated, as well as the tags in the various `Dockerfile`, and `output.go`.

**Comment by deitch** [comment]
CI green again. Awaiting another maintainer's approval and will merge.

**Comment by deitch** [comment]
> Thanks for chasing this

I think we might be able to patch the one missing piece - `linuxefi` - from coreos into upstream master. It would be better than all we are doing here. Will try and find time to do over next few weeks. Depends on how hard it turns out to be.

**Pull request merged**